### PR TITLE
Properly handle end of file hex format line

### PR DESCRIPTION
**Pull request by andrewpono**
Properly handle end of file hex format line. Update the checksum calculator to the spec one.

## Description

- Properly handle end of file hex format line (:00000001FF). In the original code if this line is encountered the field type is not set, leaving it at the default (0 = Data), however Data field is null. This throws an exception down the line when this line's Data is accessed by the caller. 
- The checksum calculation appears to be non-spec, replaced with the proper one. The old one does not fail, but may not detect the mismatch. Update: feel free to disregard, it is just a clever algorithm that went above my head.

## Motivation and Context

- Fixes hex conversion exception.

## How Has This Been Tested?

Tested with [this file](https://github.com/yhyuan/STM32-Examples/blob/4eb524553291cbebbf5c5f95cec8f611f8311d6c/15-M3-SPI%20(W25X16-W25Q16)/%E2%89%A5%C3%83%E2%80%93%C3%9A/Output/STM32-DEMO.hex), other STM32 targeted hex files.

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [x] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [ ] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [x] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [ ] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [ ] I have added new tests to cover my changes.

### CODE REVIEW

File: source/HexFormat.cs [hunk 1]
```diff
@@ -55,21 +55,16 @@ public HexFormat(string hexLine)
...
-                Crc = data[4];
-                IsValidRecord = Crc == 0xFF;
-            }
+            IsValidRecord = CheckCrc(data);
+
+            NumberOfBytes = data[0];
+            Address = BinaryPrimitives.ReadUInt16BigEndian(data.AsSpan().Slice(1, 2));
+            HexFieldType = (HexFieldType)data[3];
+            Data = new byte[NumberOfBytes];            
+            if (NumberOfBytes > 0) // no data bytes in the end of file line.
```
Comment (josesimoes) [inline source/HexFormat.cs]: Please open/close brackets on `if`. 
Comments on the line above the code.

File: source/HexFormat.cs [hunk 2]
```diff
@@ -91,12 +86,13 @@ private byte[] ConvertHex2Bin(string input)
 
         private bool CheckCrc(byte[] data)
         {
-            var crc = 0;
-            for (int i = 0; i < data.Length; i++)
+            int crc = 0;
+            for (int i = 0; i < data.Length - 1; i++) // crc is calculated for all data but the last byte (checksum)
```
Comment (josesimoes) [inline source/HexFormat.cs]: Comments on the line above the code.

**Review by josesimoes — COMMENTED** [review]
Interesting... thanks for taking the time to fix these. 
Just a couple of minor code style comments.

**Comment by josesimoes** [comment]
@andrewpono there are unit tests failing. Can you address that please? 
Also, if you could add (or update) Unit Test with your file to cover the changes you've made that would be perfect.

**Comment by andrewpono** [comment]
Both unit tests I've just changed were actually enabling the problem in the first place: no check for field type let Data type to slip by, null Data was then accessed by the caller mistaking the end of file line for data line.

**Review by josesimoes — APPROVED** [review]
LGTM! 👍🏻

**Comment by andrewpono** [comment]
Thanks for promptly merging! This is not the only problem with the code however. The converter always assumes contiguous memory map in the source hex file and discards memory start address, so this will only work if the start address is 0, which is almost never true. The changes to fix that are much more extensive, not sure I can produce code of suitable quality. I'll try.

**Comment by josesimoes** [comment]
@andrewpono this tool was added for internal consumption in our pipelines. Actually, since the release of ST programmer by ST, there is no more need for this tool as it can handle the BIN and HEX formats to program STM32 MCUs through DFU.
All this to say that it's doomed to be labeled as discontinued and it will be archived.

Of course, if it's of use to you, feel free to keep improving it.

**Comment by andrewpono** [comment]
I know Cube Programmer tools are no longer accepting dfu format, but since there are 2 dfu bootloader drivers out there in the wild I can't assume the user has the newer WinUSB driver installed, nor am I willing to force the user to update the driver. This is specific to my situation. So I have to go through either driver, which means using legacy DfuSe tools, which accept dfu format only. As for the hex format conversion, I was not using that for a while, bin is my primary, but later decided to test it to make it compatible with Cube Programmer API. This is how I found out. Anyway, the fix is pretty much done, I just have to test it.

**Pull request merged**